polar-core/src/numerics.rs

```diff
@@ -61,9 +66,67 @@ impl PartialEq for Numeric {
     }
 }
 
+impl Eq for Numeric {}
+
 /// There are 53 bits of mantissa in an IEEE 754 double precision float.
 const MOST_POSITIVE_EXACT_FLOAT: i64 = 1 << 53;
 
+/// -i64::MIN is 2**63. The maximum positive i64 is 2**63 - 1, but this
+/// isn't representable as a double. So, we first cast i64::MIN to f64
+/// then flip the sign to get 2 ** 63.
+const MOST_POSITIVE_I64_FLOAT: f64 = -(i64::MIN as f64);
+const MOST_NEGATIVE_I64_FLOAT: f64 = i64::MIN as f64;
+
+impl Hash for Numeric {
+    fn hash<H>(&self, state: &mut H)
+    where
+        H: Hasher,
+    {
+        match self {
+            Numeric::Integer(i) => {
+                discriminant(self).hash(state);
+                *i as u64
+            }
+            Numeric::Float(f) => match f.classify() {
+                FpCategory::Zero => {
+                    // Canonicalize zero representations.
+                    discriminant(&Numeric::Integer(0)).hash(state);
+                    0u64
+                }
+                FpCategory::Nan => {
+                    // Randomize NaN hashes so they always miss.
```

Comment:
This is raising a bunch of red flags.

Reply:
Here's my rationale: `NaN != NaN`, for any `NaN`, and therefore `hash(NaN) != hash(NaN)` for any `NaN`. The only alternative I've thought of is to treat `NaN` as non-ground, but this seems simpler and cleaner, if a little tricky.

Reply:
(Repeating what I said in person): I think it's okay for NaNs to hash equals to each other, and for us to treat rule filtering as not being a strict process. I.e. it can let more things through than it should. We will still unify args/params when selecting a rule so it doesn't invalidate that.

Alternatively/additionally, marking NaN as not ground I think would be a good idea.

Personally I disagree its simpler/cleaner: introduces a new dependency which relies on OS features == potentially more trouble. And it makes hashing non-deterministic which is sort of worrying.

Reply:
Abandoned.